logstash-core/src/main/java/org/logstash/config/ir/CompiledPipeline.java

```diff
@@ -0,0 +1,412 @@
+
+/**
+ * <h3>Compiled Logstash Pipeline Configuration.</h3>
+ * This class represents an executable pipeline, compiled from the configured topology that is
+ * learnt from {@link PipelineIR}.
+ * Each compiled pipeline consists in graph of {@link Dataset} that represent either a {@code filter}
+ * or an {@code if} condition.
+ */
+public final class CompiledPipeline {
+
+    /**
+     * Configured inputs.
+     */
+    private final Collection<IRubyObject> inputs;
+
+    /**
+     * Configured Filters, index by their ID as returned by {@link PluginVertex#getId()}.
+     */
+    private final Map<String, RubyIntegration.Filter> filters;
+
+    /**
+     * Immutable collection of filters that flush on shutdown.
+     */
+    private final Collection<RubyIntegration.Filter> shutdownFlushes;
+
+    /**
+     * Immutable collection of filters that flush periodically.
+     */
+    private final Collection<RubyIntegration.Filter> periodicFlushes;
+
+    /**
+     * Configured outputs.
+     */
+    private final Map<String, RubyIntegration.Output> outputs;
+
+    /**
+     * Parsed pipeline configuration graph.
+     */
+    private final PipelineIR graph;
+
+    /**
+     * Ruby pipeline object.
+     */
+    private final RubyIntegration.Pipeline pipeline;
+
+    public CompiledPipeline(final PipelineIR graph, final RubyIntegration.Pipeline pipeline) {
+        this.graph = graph;
+        this.pipeline = pipeline;
+        inputs = setupInputs();
+        filters = setupFilters();
+        outputs = setupOutputs();
+        shutdownFlushes = Collections.unmodifiableList(
+            filters.values().stream().filter(RubyIntegration.Filter::hasFlush)
+                .collect(Collectors.toList())
+        );
+        periodicFlushes = Collections.unmodifiableList(
+            shutdownFlushes.stream().filter(RubyIntegration.Filter::periodicFlush)
+                .collect(Collectors.toList())
+        );
+    }
+
+    public Collection<RubyIntegration.Filter> shutdownFlushers() {
+        return shutdownFlushes;
+    }
+
+    public Collection<RubyIntegration.Filter> periodicFlushers() {
+        return periodicFlushes;
+    }
+
+    public Collection<RubyIntegration.Output> outputs() {
+        return Collections.unmodifiableCollection(outputs.values());
+    }
+
+    public Collection<RubyIntegration.Filter> filters() {
+        return Collections.unmodifiableCollection(filters.values());
+    }
+
+    public Collection<IRubyObject> inputs() {
+        return inputs;
+    }
+
+    public RubyIntegration.Plugin registerPlugin(final RubyIntegration.Plugin plugin) {
+        plugin.register();
+        return plugin;
+    }
+
+    /**
+     * This method contains the actual compilation of the {@link Dataset} representing the
+     * underlying pipeline from the Queue to the outputs.
+     * @return Compiled {@link Dataset} representation of the underlying {@link PipelineIR} topology
+     */
+    public Dataset buildFilterFunc() {
```

Comment:
the pipeline contains more than just filters (outputs, too!), so calling this "filter" seems like a bad name.